File: WoofIOS/SwiftPackages/Data/DataSource/Sources/DataSourceFakes/Dogs/FakeDogsDataSource.swift
import Combine
import DI
import DTO
import DataSource
import Foundation
import Utils

@MockApi
public final class FakeDogsDataSource: DogsDataSourceImplementing {
    public var dogs: [DogDTO] = []
    public var getDogsError: DataSourceError?
    public var getDogsCount = 0
    public var getDogProfileError: DataSourceError?
    public var getDogProfileCount = 0

    private let scheduler: SchedulerProviding

    public func getDogs(offset: Int, limit: Int) -> AnyPublisher<DogsPageDTO, DataSourceError> {
        Deferred { [weak self] () -> AnyPublisher<DogsPageDTO, DataSourceError> in
            guard let self else { return Empty().eraseToAnyPublisher() }
            getDogsCount += 1
            if let getDogsError {
                return Fail(error: getDogsError).eraseToAnyPublisher()
            }
            return Just(
                DogsPageDTO(results: Array(dogs.dropFirst(offset).prefix(limit)), offset: offset, total: dogs.count)
            )
            .setFailureType(to: DataSourceError.self)
            .eraseToAnyPublisher()
        }
        .receive(on: scheduler.mainScheduler)
        .eraseToAnyPublisher()
    }

    public func getDogProfile(dogId: Int) -> AnyPublisher<DogProfileDTO, DataSourceError> {
        Deferred { [weak self] () -> AnyPublisher<DogProfileDTO, DataSourceError> in
            guard let self else { return Empty().eraseToAnyPublisher() }
            getDogProfileCount += 1
            if let getDogProfileError {
                return Fail(error: getDogProfileError).eraseToAnyPublisher()
            }
            guard let dog = dogs.first(where: { $0.id == dogId }) else {
                return Fail(error: DataSourceError.notFound(dogId: dogId)).eraseToAnyPublisher()
            }
            return Just(profileOf(dog)).setFailureType(to: DataSourceError.self).eraseToAnyPublisher()
        }
        .receive(on: scheduler.mainScheduler)
        .eraseToAnyPublisher()
    }

    private func profileOf(_ dog: DogDTO) -> DogProfileDTO {
        DogProfileDTO(
            dog: dog,
            breed: "Golden Retriever",
            ageInYears: 4,
            size: DogSizeDTO.large,
            neighborhood: "Kolonaki, Athens",
            personality: ["Playful", "Cuddly"],
            favoriteActivity: "Swimming at the beach",
            bio: "Loves swimming, tennis balls, and stealing snacks."
        )
    }
}

File: WoofAndroid/data/datasource/src/testFixtures/kotlin/com/perrystreet/woof/datasource/dogs/FakeDogsDataSource.kt
package com.perrystreet.woof.datasource.dogs

import com.perrystreet.woof.dto.dog.DogDTO
import com.perrystreet.woof.dto.dog.DogProfileDTO
import com.perrystreet.woof.dto.dog.DogSizeDTO
import com.perrystreet.woof.dto.dog.DogsPageDTO
import io.reactivex.rxjava3.core.Single

@org.koin.core.annotation.Single
class FakeDogsDataSource : IDogsDataSource {
    var dogs: List<DogDTO> = emptyList()
    var getDogsError: Throwable? = null
    var getDogsCount: Int = 0
    var getDogProfileError: Throwable? = null
    var getDogProfileCount: Int = 0

    override fun getDogs(offset: Int, limit: Int): Single<DogsPageDTO> {
        getDogsCount++
        return getDogsError?.let { Single.error(it) }
            ?: Single.just(
                DogsPageDTO(results = dogs.drop(offset).take(limit), offset = offset, total = dogs.size),
            )
    }

    override fun getDogProfile(dogId: Long): Single<DogProfileDTO> {
        getDogProfileCount++
        return getDogProfileError?.let { Single.error(it) }
            ?: Single.just(profileOf(dogs.first { it.id == dogId }))
    }

    private fun profileOf(dog: DogDTO) = DogProfileDTO(
        dog = dog,
        breed = "Golden Retriever",
        ageInYears = 4,
        size = DogSizeDTO.Large,
        neighborhood = "Kolonaki, Athens",
        personality = listOf("Playful", "Cuddly"),
        favoriteActivity = "Swimming at the beach",
        bio = "Loves swimming, tennis balls, and stealing snacks.",
    )
}
